### DISCUSSION TITLE
Welcome to scryber.core discussions

### DISCUSSION THREAD

**Discussion by richard-scryber** [richard-scryber/scryber.core#32] [General]
## 👋 Welcome!
  * We would love to hear about what you are doing with scryber
  * Are there any things you would like to see.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with scryber.

**Comment by DaveDansey** [comment]
Hi,

My name's Dave and I've been coding since the ZX81 spectrum days :-)  I've also been hiking and climbing mountains for as long!  I'm from Hertfordshire in the UK (so can't wait until lockdown is over and I can get back to Wales).

I've been looking for a simple free PDF creation library recently, with a view to programatically creating some PDF's of mountain\trig point stats from a simple app I'm creating for my own use.

Yours is awesome!  I've created a Document and added a Section with contents (simple Div's and Tables) and successfully rendered it to a pdf file.

I feel I must be missing something obvious but I can't work out how to add content to the Header or Footer (or ContinuationHeader \ ContinuationFooter) properties of a Section.  The Section.Contents is a ComponentList I can add Div's and Table's to, but the header\footer properties are null and of type IPDFTemplate.

I've tried to set Section.Header to various objects but can't find any that implement IPDFTemplate... I've also tried to add various objects to the Contents collection (like Head1, PDFPageHeader) - thinking that certain types of content added to the main collection may end up in the header or footer.  Nothing I've tried so far works.

I'm just trying to create a basic section header and a footer which repeats on all pages containing the page number.

Any help would be much appreciated,
Many thanks,

Dave

**Reply by richard-scryber** [reply]
Hi Dave, unless you can convince me you had the 16k expansion pack... I might struggle ;-)

Thank you for the positive feedback - really appreciate it!

Are you using an xml template? There is a &lt;pdf:PageNumber /&gt; element you can add to your content to show the current page number (https://scrybercore.readthedocs.io/en/v1.0.0.20-beta/document_pagenumbers.html)

If you are using html templates then we have added support for a &lt;page /&gt; tag in html. (working on the documentation)

However if you are using code directly, then there is the Scryber.Data.ParsableTemplateGenarator which will need the contents and namespaces used. But I would definitely recommend trying to go with the XML/HTML first.

If this is of no help whatsoever, just put some examples in, and I think we can figure a solution.

**Reply by DaveDansey** [reply]
Oh yes, the little black box that plugged into the back, but was such a rubbish connection that if you knocked it even slightly whist turned on, the whole PC would crash.  We put a strip of blue-tac just above the connector which helped greatly... can't have it crashing when in the middle of an important bat & ball game!

I'm just trying to create the most basic document in code at the minute, not using templates... I was thinking of potentially exploring them later on if time allows, but I don't really need any styling, just plain text headings and some simple tables generated directly from code (by just iterating a data table and dumping some info into the pdf table).

```
        static void Main(string[] args)
        {
            // Create a new document and section.
            Document doc = new Document();
            Section section = new Section();

            // Add some content to the section body.
            section.Contents.Add(new Label() { Text = "This is a label in the document body" });


            // How to add Header content...?
            section.Header = null;   // This property is null and of type IPDFTemplate... unable to add content to the header in the same way I can add content to the 'body' (e.g. section.Contents)


            // How to add ContinunationFooter content...?
            section.ContinuationFooter = null;  // As above, I can't work out how to create a footer in code, containing the page number.


            // Create a file stream to write the file to.
            string fileName = @"C:\Temp\ScryberTest.pdf";
            Stream pdfStream = new FileStream(fileName, FileMode.Create);

            // Add section to document pages collection.
            doc.Pages.Add(section);

            // Save pdf to file.
            doc.InitializeAndLoad();
            doc.RenderToPDF(pdfStream);
        }
```

The above creates a document with a single label in its "body".  I've also created a simple function which can create a dummy table with n rows, so have managed to create a section and add to it a table with hundreds of rows, which wraps over lots of pages.

It's easy to add to the section.Contents collection above... I can add new Labels, Divs, Tables and probably lots other things like images.

I was expecting\hoping that the Header\Footer properties would be the same, so I could just add labels, tables, or whatever content I wanted to appear in the header or footer.

I realise the documentation is geared towards using xml and templates as that's how most people would use it... but under the heading "To code or not to code…" it does state "...you are simply creating the same as you could in code." - so I assume it's possible to set the header and footer somehow when using code-only, but I haven't figured out how.

Cheers,
Dave

**Reply by richard-scryber** [reply]
Hi Dave

I did a quick sample, using either the the Scryber.Data.ParsableTemplateGenerator that we use internally when parsing XML, or a separate class, 'MyFooter', implementing the IPDFTemplate (there is only one method)


     public void HelloWorld()
     {
            var path = System.Environment.CurrentDirectory;
            path = System.IO.Path.Combine(path, "../../../Content/HTML/HelloWorld.html");

            using (var doc = Document.ParseDocument(path))
            {
                var section = new Scryber.Components.Section();

                //xml string content
                var head = new Scryber.Data.ParsableTemplateGenerator();
                head.XmlContent = "<doc:Div xmlns:doc='http://www.scryber.co.uk/schemas/core/release/v1/Scryber.Components.xsd'><doc:PageNumber /></doc:Div>";
                section.Header = head;

                //class implementation
                var foot = new MyFooter();
                section.Footer = foot;

                doc.Pages.Add(section);

                using (var stream = GetOutputStream("HelloWorld.pdf"))
                {
                    doc.SaveAsPDF(stream);
                }

            }
       }

        private class MyFooter : IPDFTemplate
        {
            public IEnumerable<IPDFComponent> Instantiate(int index, IPDFComponent owner)
            {
                var div = new Div();
                div.Contents.Add(new PageNumberLabel());
                return new IPDFComponent[] { div };
            }
        }

I hope this helps.

**Reply by richard-scryber** [reply]
The reason they are not the same as the content, is that they may be created multiple times (with different content) on each page. So the templating mechanism is used each time an instance is needed - new page new instance.

**Reply by DaveDansey** [reply]
Hi Richard,

Thank you, that's perfect!  So obvious now you show how of course, I just hadn't worked out that I needed to create a new class which inherits from IPDFTemplate.

It's just what I want for now.  When I get time I'll definitely try creating some templates - good to know you can you can combine a mix of templates and code, really flexible and probably the way to go if you want some more complex layouts and styling combined with some custom logic and data inserted at run-time in various ways.

Thanks again for the assistance,
Dave

**Reply by richard-scryber** [reply]
Glad it works for you Dave. And I agree on the code vs declarative. Whatever suits is right.

If it is a big table, one of the things I was asked previously was to improve the performance, and we added a DataStyleIdentifier property to effectively cache the full style rather than build each time.

               TableCell tblcell = new TableCell();
                tblcell.DataStyleIdentifier = "AnyString";

You could try this on your rows and cells and may find it works for you. I'd be interested to know the results.

Have fun with your documents - That is why we do this, I think.

**Reply by DaveDansey** [reply]
Good to know, thanks, I'll definitely check out that new property.  I've just done the bare minimum of getting the data on the page so far but would be nice to add some styling and remove some of the inner border lines at some point :-)

Yeah I've not lost the 'having fun' aspect of coding over the years!... I got into it originally as my father was a programmer (machine code and assembly language etc), but always liked the way it could make a manual process much more efficient.  When planning a hike \ mountain climb I currently gather quite a bit of info from various places (maps, routes, trig\mountain info, weather etc).  The dream would be automating a lot of this, so I can just press a few buttons and gather all info into a pdf ready to print out and take with me.

**Reply by richard-scryber** [reply]
How did you get on Dave? Was the library able to do what you needed?
Always looking for feedback, as it’s the only way we can make it better.

**Reply by DaveDansey** [reply]
Hi Richard,

It's worked great thank you, I've got tables of data populated and page headers\footers all sorted and being generated in code.

It's very basic and could do with styling to improve the look, but it's just a hobby-thing for personal use and unfortunately the day job and other commitments don't leave much time to invest in fun things :-)

I'm sure I'll have further questions when I can get back to it (rather busy at present) - I'll post a new thread when I do.

When you have time, a few code-based examples in the documentation may help others like me, specifically how to generate the headers\footers (by creating a class which inherits IPDFTemplate).  Apart from that, all I've read about and used so far is really intuitive and easy to understand, so nice one, much appreciated!

Dave

**Comment by wbijker** [comment]
Hi, Willem from sunny South Africa here. Started coding in primary school and never stop since. Currently part of a small startup. 

I've explored and used various tools and utilities for creating PDF's in core - most of them wrappers around wkhtmltopdf. Even built a library myself ([itechoPdf](https://github.com/itechodev/itechopdf/)) which did things a little faster and better than the others. And the library was generally working great in spite of many caveats and limitations. Generation time and large file size became a huge problem when I start generating 100+ pages. 

So, I searched for something more "native" or a more programatic approach to build PDF's - given that they are usually faster.  And I found Scryber. Which I found to be brilliant. Got me a 200x performance increase and a 40x smaller file size!

One thing missing before I can fully move to Scryber: vertical text. I had a quick read and looks like rotations / writing-mode is not implemented in the PDF spec? 

Keep up your good work. 

PS I will not be surprised if Scryber is found at the very top of the best C# PDF generation library list.

Willem

**Reply by richard-scryber** [reply]
Hi Wllem, firstly - thank you for the kind words. I hope you are right :-)

For vertical text, the underlying code does support transformations (rotations etc.), if that is sufficient for you.
I can probably get something out working in the next beta release fairly quickly. It's just not been fully tested yet.

It can however be applied in the document object model. For example

Just a div with some text in in a template.

`<div id="mydiv" style="color:blue;  background-color: orange; padding: 5px;
                    top: 50px; left: 100px; height: 100px; width:200px;
                    vertical-align:middle; text-align:center;" >This is some text</div>`

And in your generation backend code...

`if(doc.TryFindAComponentByID("mydiv", out div))
                {
                    div.Style.SetValue(StyleKeys.TransformRotateKey, 90);
                }`

And you should get an output similar to below...

![image](https://user-images.githubusercontent.com/65354830/116982013-33577500-acc0-11eb-8918-e878a612381f.png)

It might need a bit of playing around to get the right positioning. 


Actual vertical text writing mode, honestly it is not something I've ever been asked about. I did look into RtL text rendering but that got into a whole heap of ligatures and layout issues.

Let me know if this helps.

Kr

Richard

**Reply by richard-scryber** [reply]
And good luck with the startup too - exciting times!

**Reply by richard-scryber** [reply]
Hi Willem - Do you mind if I use your quote around performance on our new [scryber](https://www.scryber.co.uk) site? We are just setting it up but a great quote.

**Reply by wbijker** [reply]
Richard, of course!  I can also present you with my performance comparisons between scryber and wkhtmltopdf if you like.

**Reply by richard-scryber** [reply]
That would be amazing! Yes please. Thanks Willem.

**Comment by tim-biasoft** [comment]
Hi, my name is Tim and I live in Hungary not too far from Budapest. I'm an independent contractor currently working on a billing system, which means PDF invoices of course!

Thank you for creating and releasing this library. Clearly a lot of thought, time and effort has gone into it and I agree with Willem that Scryber will be top of the list of best C# PDF generation libraries.

If it's not a delicate subject, can I ask if there is a commercial side to this project?

Tim

**Reply by richard-scryber** [reply]
Hi Tim - **Love Budapest!** Took a train trip there from the UK a few years ago via Paris and Prague. What a different world that was then! But awesome break and waking up on the overnighter, travelling along the side of the Danube - special.

Thanks for the support and vote of confidence. I think it's a good engine and glad you and others are thinking so too. 

I do have an opportunity I am taking now, to look at options for growing scryber. It's not a delicate subject at all.
For me it is really important that it will always stay open source. There are many other revenue streams to look at with sponsoring, courses, training, videos that can create income. Growing and assisting the 'scribbler' community is key to this.

I believe the business world and developer community has moved on from large library subscription costs, and monolithic platforms. There are better ways. And we can grow support and possibly a SaaS platform to balance the open engine.

If you think of something, just say :-)

I'm also keen to know from others where is best for social updates, mentions, quick hints, general chat. I like Reddit - feels more considered in its content. But really trying to understand where others would like this sort of info and where they may want to contribute to.

**Reply by tim-biasoft** [reply]
I found https://www.scryber.online/ which I guess is related, looks like you've got things covered on the business side of things :D  
One thing I didn't see mentioned was a bespoke service to turn client designs into templates, might be a way to get revenue from the open source library users too.  

I'm not big on social media so I can't help you with that one!

**Reply by richard-scryber** [reply]
I can see where you are thinking Tim, and I can really see that potential. 

It is a great service to offer but for personal reasons I’d rather focus on the core capabilities and leave others to build the bespoke. 

If I could do anything to help promote their services though - I would. Be nice to showcase.

**Comment by alikhaaan** [comment]
I want to add Slider in the PDF. Is it possible to add SVG Slider or Can i add Slider using Javascript?

**Reply by alikhaaan** [reply]
@richard-scryber

**Reply by richard-scryber** [reply]
Hi - Sorry for the slow response, I know it would be possible to create an interactive slider in the pdf, but using the OOB components we have now - I'm not sure where you would start. It is possible to write and include tour own custom components, and extend the functionality, but it would be a significant undertaking. I'd love to know if you were able, either using SVG or your own components.

I can add it to our list as forms is something we need to do in the future, but I'm a way off that yet.

Richard

**Comment by samaravathi** [comment]
Hi Richard,

Thanks for this great repo.
Is there a way to get entire html after all the data is bound to the xhtml template?
I want to save just the html.

Regards

**Reply by richard-scryber** [reply]
Hi Samaravathi

It is something I've looked at for extracting the ePub format, rather than PDF. And generating the html from the components itself would be fairly trivial with overriding methods...

```
public override void ToHtml(StreamWriter sw){
    sw.Append("<table ");
    this.AddHtmlAttributes(sw);
    sw.Append(">");
    this.ChildrenToHtml(this.Rows, sw);
    sw.Append("</table>");
}
```

From the library perspective, there would be a number of configuration options that would need to be included, inline css, formatting, xml or string, etc. that takes it beyond the trivial and a more intensive sprint of effort. I have now added it to the ever growing backlog. I would happily take your input on the formatting options, but it might be that your timeframes mean it is easier to add your own implementation that fits your needs.

You might also want to look at other libraries and solutions out there to see if they work for you.
[https://github.com/Handlebars-Net/Handlebars.Net](https://github.com/Handlebars-Net/Handlebars.Net) came up in a brief search.

Richard
